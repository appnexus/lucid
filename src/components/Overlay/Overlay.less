@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.lucid-Overlay {
	.transition-group-animation-fade();
	display: flex;
	justify-content: center;
	align-items: center;
	position: fixed;
	left: 0;
	top: 0;
	width: 100vw;
	height: 100vh;
	overflow: hidden;
	z-index: @zindex-modal;
	background-color: @color-gray-60;

	&&-is-not-modal {
		background-color: @color-transparent;
		// Setting this to `none` allows us to use flex box to center the content
		// on the screen while still allowing the user to interact with elements
		// behind it
		pointer-events: none;

		// Have to set the pointer events back to initial so the user can interact
		// with the contents
		> * {
			pointer-events: auto;
		}
	}
}
