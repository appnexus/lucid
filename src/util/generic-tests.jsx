import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
