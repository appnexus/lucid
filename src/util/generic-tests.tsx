import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
