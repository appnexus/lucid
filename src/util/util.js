import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
