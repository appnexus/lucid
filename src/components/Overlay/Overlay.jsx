import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Portal from '../Portal/Portal';
import ReactTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../util/component-types';

const cx = lucidClassNames.bind('&-Overlay');

const { string, bool, func, node } = PropTypes;

/**
 * {"categories": ["utility"], "madeFrom": ["Portal"]}
 *
 * Overlay is used to block user interaction with the rest of the app until they
 * have completed something.
 */
const Overlay = createReactClass({
	displayName: 'Overlay',

	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root element.
		 */
		className: string,

		/**
		 * Generally you should only have a single child element so the centering
		 * works correctly.
		 */
		children: node,

		/**
		 * Controls visibility.
		 */
		isShown: bool,

		/**
		 * Determines if it shows with a gray background. If `false`, the
		 * background will be rendered but will be invisible, except for the
		 * contents, and it won't capture any of the user click events.
		 */
		isModal: bool,

		/**
		 * Set your own id for the `Portal` is that is opened up to contain the
		 * contents. In practice you should never need to set this manually.
		 */
		portalId: string,

		/**
		 * Fired when the user hits escape.
		 *
		 * Signature: `({ event, props }) => {}`
		 */
		onEscape: func,

		/**
		 * Fired when the user clicks on the background, this may or may not be
		 * visible depending on `isModal`.
		 *
		 * Signature: `({ event, props }) => {}`
		 */
		onBackgroundClick: func,
	},

	getDefaultProps() {
		return {
			isShown: false,
			isModal: true,
			onEscape: _.noop,
			onBackgroundClick: _.noop,
		};
	},

	getInitialState() {
		return {
			// This must be in state because getDefaultProps only runs once per
			// component import which causes collisions
			portalId: this.props.portalId || _.uniqueId('Overlay-Portal-'),
		};
	},

	componentDidMount() {
		if (window && window.document) {
			window.document.addEventListener('keydown', this.handleDocumentKeyDown);
		}
	},

	componentWillUnmount() {
		if (window && window.document) {
			window.document.removeEventListener(
				'keydown',
				this.handleDocumentKeyDown
			);
		}
	},

	handleDocumentKeyDown(event) {
		// If the user hits the "escape" key, then fire an `onEscape`
		// TODO: use key helpers
		if (event.keyCode === 27) {
			this.props.onEscape({ event, props: this.props });
		}
	},

	handleDivRef(divDOMNode) {
		// Store the dom node so we can check if it's clicked on later
		this._divDOMNode = ReactDOM.findDOMNode(divDOMNode);
	},

	handleBackgroundClick(event) {
		// Use the reference we previously stored from the `ref` to check what
		// element was clicked on.
		if (this._divDOMNode && event.target === this._divDOMNode) {
			this.props.onBackgroundClick({ event, props: this.props });
		}
	},

	render() {
		const {
			className,
			isShown,
			isModal,
			children,
			...passThroughs
		} = this.props;

		const { portalId } = this.state;

		return (
			<Portal portalId={portalId}>
				<ReactTransitionGroup
					transitionName={cx('&')}
					transitionEnterTimeout={300}
					transitionLeaveTimeout={300}
				>
					{isShown
						? <div
								{...omitProps(passThroughs, Overlay)}
								className={cx(className, '&', {
									'&-is-not-modal': !isModal,
								})}
								onClick={this.handleBackgroundClick}
								ref={this.handleDivRef}
							>
								{children}
							</div>
						: null}
				</ReactTransitionGroup>
			</Portal>
		);
	},
});

export default Overlay;
