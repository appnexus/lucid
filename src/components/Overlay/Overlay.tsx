import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Portal from '../Portal/Portal';
import { CSSTransition } from 'react-transition-group';
import { lucidClassNames, uniqueName } from '../../util/style-helpers';
import { omitProps, StandardProps } from '../../util/component-types';

const cx = lucidClassNames.bind('&-Overlay');

const { string, bool, func, node } = PropTypes;

export interface IOverlayProps
	extends StandardProps,
		React.DetailedHTMLProps<
			React.HTMLAttributes<HTMLDivElement>,
			HTMLDivElement
		> {
	/** Controls visibility. */
	isShown: boolean;

	/** Enables animated transitions during expansion and collapse. */
	isAnimated: boolean;

	/** Determines if it shows with a gray background. If `false`, the
		background will be rendered but will be invisible, except for the
		contents, and it won't capture any of the user click events. */
	isModal: boolean;

	/** Set your own id for the \`Portal\` is that is opened up to contain the
		contents. In practice you should never need to set this manually. */
	portalId?: string;

	/** Fired when the user hits escape. */
	onEscape: ({
		event,
		props,
	}: {
		event: KeyboardEvent;
		props: IOverlayProps;
	}) => void;

	/** Fired when the user clicks on the background, this may or may not be
		visible depending on \`isModal\`. */
	onBackgroundClick: ({
		event,
		props,
	}: {
		event: React.MouseEvent;
		props: IOverlayProps;
	}) => void;
}

interface IOverlayState {
	portalId: string;
}

export const defaultProps = {
	isShown: false,
	isModal: true,
	onEscape: _.noop,
	onBackgroundClick: _.noop,
	isAnimated: true,
};

class Overlay extends React.Component<IOverlayProps, IOverlayState, {}> {
	static displayName = 'Overlay';

	static peek = {
		description: `
			Overlay is used to block user interaction with the rest of the app
			until they have completed something.
		`,
		categories: ['utility'],
		madeFrom: ['Portal'],
	};

	static propTypes = {
		/**
			Appended to the component-specific class names set on the root element.
		*/
		className: string,

		/**
			Generally you should only have a single child element so the centering
			works correctly.
		*/
		children: node,

		/**
			Controls visibility.
		*/
		isShown: bool,

		isAnimated: bool,

		/**
			Determines if it shows with a gray background. If \`false\`, the
			background will be rendered but will be invisible, except for the
			contents, and it won't capture any of the user click events.
		*/
		isModal: bool,

		/**
			Set your own id for the \`Portal\` is that is opened up to contain the
			contents. In practice you should never need to set this manually.
		*/
		portalId: string,

		/**
			Fired when the user hits escape.  Signature: \`({ event, props }) => {}\`
		*/
		onEscape: func,

		/**
			Fired when the user clicks on the background, this may or may not be
			visible depending on \`isModal\`.  Signature: \`({ event, props }) => {}\`
		*/
		onBackgroundClick: func,
	};

	private rootHTMLDivElement = React.createRef<HTMLDivElement>();

	static defaultProps = defaultProps;

	state = {
		// This must be in state because getDefaultProps only runs once per
		// component import which causes collisions
		portalId: this.props.portalId || uniqueName('Overlay-Portal-'),
	};

	componentDidMount(): void {
		if (window && window.document) {
			window.document.addEventListener('keydown', this.handleDocumentKeyDown);
		}
	}

	componentWillUnmount(): void {
		if (window && window.document) {
			window.document.removeEventListener(
				'keydown',
				this.handleDocumentKeyDown
			);
		}
	}

	handleDocumentKeyDown = (event: KeyboardEvent): void => {
		// If the user hits the "escape" key, then fire an `onEscape`
		// TODO: use key helpers
		if (event.keyCode === 27) {
			this.props.onEscape({ event, props: this.props });
		}
	};

	handleBackgroundClick = (event: React.MouseEvent): void => {
		// Use the reference we previously stored from the `ref` to check what
		// element was clicked on.
		if (
			this.rootHTMLDivElement.current &&
			event.target === this.rootHTMLDivElement.current
		) {
			this.props.onBackgroundClick({ event, props: this.props });
		}
	};

	render(): React.ReactNode {
		const {
			className,
			isShown,
			isModal,
			isAnimated,
			children,
			...passThroughs
		} = this.props;

		const { portalId } = this.state;

		const overlayElement = (
			<div
				{...omitProps(passThroughs, undefined, Object.keys(Overlay.propTypes))}
				className={cx(className, '&', {
					'&-is-not-modal': !isModal,
					'&-is-animated': isAnimated,
				})}
				onClick={this.handleBackgroundClick}
				ref={this.rootHTMLDivElement}
			>
				{children}
			</div>
		);

		return (
			<Portal portalId={portalId}>
				{isAnimated ? (
					<CSSTransition
						in={isShown}
						classNames={cx('&')}
						timeout={300}
						unmountOnExit
					>
						{overlayElement}
					</CSSTransition>
				) : isShown ? (
					overlayElement
				) : null}
			</Portal>
		);
	}
}

export default Overlay;
