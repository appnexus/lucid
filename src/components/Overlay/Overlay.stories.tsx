import _ from 'lodash';
import React from 'react';
import createClass from 'create-react-class';
import Overlay from './Overlay';
import Button from '../Button/Button';

export default {
	title: 'Utility/Overlay',
	component: Overlay,
	parameters: {
		docs: {
			description: {
				component: (Overlay as any).peek.description,
			},
		},
	},
};

/* Default */
export const Default = () => {
	const Component = createClass({
		getInitialState() {
			return {
				isShown: false,
			};
		},

		handleOpenClose(isShown: any) {
			this.setState({
				isShown: isShown,
			});
		},

		render() {
			return (
				<div>
					<Button
						onClick={_.partial(this.handleOpenClose, !this.state.isShown)}
					>
						Toggle
					</Button>

					<Overlay
						isShown={this.state.isShown}
						onEscape={_.partial(this.handleOpenClose, false)}
						onBackgroundClick={_.partial(this.handleOpenClose, false)}
					>
						<div style={{ color: '#fff' }}>
							User cannot interact with the background (except scrolling).
						</div>
					</Overlay>
				</div>
			);
		},
	});

	return <Component />;
};
Default.storyName = 'Default';

/* No Modal */
export const NoModal = () => {
	const Component = createClass({
		getInitialState() {
			return {
				isShown: false,
			};
		},

		handleOpenClose(isShown: any) {
			this.setState({
				isShown: isShown,
			});
		},

		render() {
			return (
				<div>
					<Button
						onClick={_.partial(this.handleOpenClose, !this.state.isShown)}
					>
						Toggle
					</Button>

					<Overlay
						isShown={this.state.isShown}
						isModal={false}
						onEscape={_.partial(this.handleOpenClose, false)}
					>
						<div
							style={{
								backgroundColor: '#eee',
								padding: '100px',
							}}
						>
							User can still interact with the background.
							<Button onClick={_.partial(this.handleOpenClose, false)}>
								Close
							</Button>
						</div>
					</Overlay>
				</div>
			);
		},
	});

	return <Component />;
};
NoModal.storyName = 'NoModal';
