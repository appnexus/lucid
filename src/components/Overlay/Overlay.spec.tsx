import React from 'react';
import { mount, shallow } from 'enzyme';
import { common } from '../../util/generic-tests';
import sinon from 'sinon';
import assert from 'assert';
import { CSSTransition } from 'react-transition-group';

import Overlay from './Overlay';

describe('Overlay', () => {
	common(Overlay, {
		getDefaultProps: () => {
			return { isShown: true };
		},
		selectRoot: wrapper => wrapper.find('.lucid-Overlay'),
	});

	it('should render body content', () => {
		const wrapper = shallow(<Overlay isShown={true}>Flux Capacitor</Overlay>);

		assert(wrapper.contains('Flux Capacitor'));
	});

	describe('when isAnimated is false', () => {
		describe('and when isShown is true', () => {
			it('should render CSSTransition with in prop as false', () => {
				const wrapper = shallow(<Overlay isShown={false} isAnimated={true}>Flux Capacitor</Overlay>);
				assert.equal(wrapper.find(CSSTransition).props().in, false);
			});
		});
		describe('and when isShown is false', () => {
			it('should render CSSTransition with in prop as false', () => {
				const wrapper = shallow(<Overlay isShown={true} isAnimated={true}>Flux Capacitor</Overlay>);
				assert.equal(wrapper.find(CSSTransition).props().in, true);
			});
		});
	});

	describe('when isAnimated is true', () => {
		it('should not render when isShown is false', () => {
			const wrapper = shallow(<Overlay isShown={false} isAnimated={false}>Flux Capacitor</Overlay>);
			assert(!wrapper.contains('Flux Capacitor'));
		});
	});



	it('should have the correct class when isModal is false', () => {
		const wrapper = shallow(
			<Overlay isShown={true} isModal={false}>
				Nerp
			</Overlay>
		);

		assert.equal(wrapper.find('.lucid-Overlay-is-not-modal').length, 1);
	});

	it('should render with isAnimated=false', () => {
		const wrapper = shallow(<Overlay isAnimated={false}>Nerp</Overlay>);

		expect(wrapper).toMatchSnapshot();
	});
});

describe('Overlay', () => {
	let testDOMNode;

	beforeEach(() => {
		testDOMNode = document.createElement('div');
		document.body.appendChild(testDOMNode);
	});

	afterEach(() => {
		testDOMNode.parentNode.removeChild(testDOMNode);
	});

	it('should fire onBackgroundClick', () => {
		const onBackgroundClick = sinon.spy();
		const wrapper = mount(
			<Overlay
				isShown={true}
				isModal={false}
				onBackgroundClick={onBackgroundClick}
				portalId={'brolo'}
			/>,
			{
				attachTo: testDOMNode,
			}
		);

		const event = document.createEvent('MouseEvents');
		event.initMouseEvent(
			'click', //event type : click, mousedown, mouseup, mouseover, mousemove, mouseout.
			true, //canBubble
			false, //cancelable
			window, //event's AbstractView : should be window
			1, // detail : Event's mouse click count
			50, // screenX
			50, // screenY
			50, // clientX
			50, // clientY
			false, // ctrlKey
			false, // altKey
			false, // shiftKey
			false, // metaKey
			0, // button : 0 = click, 1 = middle button, 2 = right button
			null // relatedTarget : Only used with some event types (e.g. mouseover and mouseout). In other cases, pass null.
		);

		document.querySelector('#brolo .lucid-Overlay').dispatchEvent(event);

		assert(onBackgroundClick.called);
		wrapper.unmount();
	});

	it('should not fire onBackgroundClick when content is clicked', () => {
		const onBackgroundClick = sinon.spy();
		const wrapper = mount(
			<Overlay
				isShown={true}
				isModal={false}
				onBackgroundClick={onBackgroundClick}
				portalId={'regiewat'}
			>
				<div id='foo'>Nope</div>
			</Overlay>,
			{
				attachTo: testDOMNode,
			}
		);

		const event = document.createEvent('MouseEvents');
		event.initMouseEvent(
			'click', //event type : click, mousedown, mouseup, mouseover, mousemove, mouseout.
			true, //canBubble
			false, //cancelable
			window, //event's AbstractView : should be window
			1, // detail : Event's mouse click count
			50, // screenX
			50, // screenY
			50, // clientX
			50, // clientY
			false, // ctrlKey
			false, // altKey
			false, // shiftKey
			false, // metaKey
			0, // button : 0 = click, 1 = middle button, 2 = right button
			null // relatedTarget : Only used with some event types (e.g. mouseover and mouseout). In other cases, pass null.
		);

		document.querySelector('#regiewat #foo').dispatchEvent(event);

		assert(onBackgroundClick.notCalled);
		wrapper.unmount();
	});

	it('should fire the onEscape handler when escape is typed', () => {
		const onEscape = sinon.spy();
		const wrapper = mount(<Overlay isShown={true} onEscape={onEscape} />, {
			attachTo: testDOMNode,
		});

		const event = document.createEvent('Event');
		event.initEvent('keydown', true, true);

		event.keyCode = 27;

		document.dispatchEvent(event);

		assert(onEscape.called);
		wrapper.unmount();
	});
});
