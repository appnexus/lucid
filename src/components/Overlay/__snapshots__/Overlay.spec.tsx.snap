// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay [common] example testing should match snapshot(s) for basic 1`] = `
<Portal
  portalId="1-Overlay-Portal-6"
>
  <CSSTransition
    classNames="lucid-Overlay"
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="lucid-Overlay lucid-Overlay-is-animated"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        User cannot interact with the background (except scrolling).
      </div>
    </div>
  </CSSTransition>
</Portal>
`;

exports[`Overlay [common] example testing should match snapshot(s) for no-modal 1`] = `
<Portal
  portalId="1-Overlay-Portal-7"
>
  <CSSTransition
    classNames="lucid-Overlay"
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="lucid-Overlay lucid-Overlay-is-not-modal lucid-Overlay-is-animated"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "backgroundColor": "#eee",
            "padding": "100px",
          }
        }
      >
        User can still interact with the background.
        <Button
          hasOnlyIcon={false}
          isActive={false}
          isDisabled={false}
          onClick={[Function]}
          type="button"
        >
          Close
        </Button>
      </div>
    </div>
  </CSSTransition>
</Portal>
`;

exports[`Overlay should render with isAnimated=false 1`] = `
<Portal
  portalId="1-Overlay-Portal-13"
/>
`;
